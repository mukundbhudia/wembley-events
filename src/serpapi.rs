use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiSearchMetaData {
    pub created_at: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiDate {
    pub start_date: String,
    pub when: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiVenue {
    pub name: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiEvent {
    pub title: String,
    pub description: String,
    pub date: SerpapiDate,
    pub link: String,
    pub venue: SerpapiVenue,
    pub thumbnail: String,
}

#[derive(Deserialize, Serialize, Default, Debug)]
pub struct SerpapiEvents {
    pub search_metadata: SerpapiSearchMetaData,
    pub events_results: Vec<SerpapiEvent>,
}

#[cfg(test)]
mod tests {
    use crate::{
        serpapi_test_output_json_1, serpapi_test_output_json_2,
        test_files::{
            serpapi_test_output_json_3, serpapi_test_output_json_4,
            serpapi_test_output_json_5_missing_date, serpapi_test_output_json_6_one_bad_json_event,
        },
        WembleyEvents,
    };

    #[test]
    fn basic_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_1();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);

        assert_eq!(wembley_events.events.len(), 10);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn empty_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_2();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);

        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn no_json_response_serpapi_events_test() {
        let mock_serpapi_response = String::new();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn no_events_in_response_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_3();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn bad_json_in_response_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_4();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn one_good_one_bad_event_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_5_missing_date();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }

    #[test]
    fn one_good_one_bad_json_event_serpapi_events_test() {
        let mock_serpapi_response = serpapi_test_output_json_6_one_bad_json_event();

        let wembley_events = WembleyEvents::new().build_events_from_html(mock_serpapi_response);
        assert_eq!(wembley_events.events.len(), 0);
        insta::assert_debug_snapshot!(wembley_events);
    }
}
