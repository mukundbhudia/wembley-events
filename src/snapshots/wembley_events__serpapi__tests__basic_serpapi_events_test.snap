---
source: src/serpapi.rs
assertion_line: 50
expression: wembley_events

---
WembleyEvents {
    events: {
        0: WembleyEvent {
            date: "19 Jun 2022",
            time: "Sun, Jun 19",
            place: "Club Wembley",
            title: "Harry Styles",
            description: "Multi-platinum recording artist Harry Styles announces the rescheduled dates for his world tour, in addition to new shows added across the globe. Styles will kick off his colossal 32-city outing...",
        },
        1: WembleyEvent {
            date: "12 Jun 2022",
            time: "Sun, 2 PM",
            place: "Wembley Stadium",
            title: "David Guetta @ Wembley Stadium",
            description: "Capital's Summertime Ball",
        },
        2: WembleyEvent {
            date: "5 Jun 2022",
            time: "Tomorrow, 1 – 10 PM",
            place: "BOXPARK Wembley",
            title: "Soulful Sundays",
            description: "Join BOXPARK Wembley and their resident DJ's for all things soulful",
        },
        3: WembleyEvent {
            date: "29 Jun 2022",
            time: "Wed, 4:30 – 8:30 PM",
            place: "Wembley Stadium",
            title: "Ed Sheeran @ Wembley Stadium",
            description: "span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...",
        },
        4: WembleyEvent {
            date: "25 Jun 2022",
            time: "Sat, 5 – 9 PM",
            place: "Wembley Stadium",
            title: "Dylan @ Wembley Stadium",
            description: "span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...",
        },
        5: WembleyEvent {
            date: "16 Aug 2022",
            time: "Tue, 5 – 9 PM",
            place: "Wembley Stadium",
            title: "Coldplay @ Wembley Stadium",
            description: "Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...",
        },
        6: WembleyEvent {
            date: "24 Jun 2022",
            time: "Fri, 5 – 9 PM",
            place: "Wembley Stadium",
            title: "Ed Sheeran + - = ÷ x Tour",
            description: "Ed Sheeran Seated Ticket & Hotel Experience This Experience Includes: • Reserved Seated Ticket (Block 104 or 141) Tickets seated together • One night hotel stay including complimentary breakfast •...",
        },
        7: WembleyEvent {
            date: "20 Aug 2022",
            time: "Sat, 5 – 9 PM",
            place: "Wembley Stadium",
            title: "Coldplay: Music of the Spheres World Tour",
            description: "Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...",
        },
        8: WembleyEvent {
            date: "26 Jun 2022",
            time: "Sun, 7 – 9 PM",
            place: "OVO Arena",
            title: "Jubin Nautiyal - RESCHEDULED",
            description: "Jubin Nautiyal at OVO Arena Wembley in Wembley Park",
        },
        9: WembleyEvent {
            date: "10 Jun 2022",
            time: "Fri, 7:30 – 11:00 PM",
            place: "The SSE Arena, Wembley",
            title: "50 Cent",
            description: "SJM Concerts 50 CENT FRI, 10 JUN 2022 at 07:30PM BST Doors Open: 06:00PM OnSale: Fri, 28 Jan 2022 at 09:30AM GMT Announcement: Wed, 26 Jan 2022 at 10:00AM GMT Hip Hop heavyweight 50 Cent, has...",
        },
    },
}
